src/common/contracts/ant-versions.ts

```diff
@@ -0,0 +1,127 @@
+  isProcessIdConfiguration,
+} from '../../types/io.js';
+import { AoSigner } from '../../types/token.js';
+import { createAoSigner } from '../../utils/ao.js';
+import { pruneTags } from '../../utils/arweave.js';
+import { InvalidContractConfigurationError } from '../error.js';
+import { AOProcess } from './ao-process.js';
+
+type ANTVersionsNoSigner = ProcessConfiguration;
+type ANTVersionsWithSigner = WithSigner<ProcessConfiguration>;
+type ANTVersionsConfig = ANTVersionsNoSigner | ANTVersionsWithSigner;
+
+export class ANTVersions {
+  // by default give read
+  static init(): AoANTVersionsRead;
+
+  // no signer give read
+  static init(config: ANTVersionsNoSigner): AoANTVersionsRead;
+
+  // with signer give write
+  static init(config: ANTVersionsWithSigner): AoANTVersionsWrite;
+
+  static init(
+    config?: ANTVersionsConfig,
+  ): AoANTVersionsRead | AoANTVersionsWrite {
+    if (config !== undefined && 'signer' in config) {
+      return new ANTVersionsWritable(config);
+    }
+    return new ANTVersionsReadable(config);
+  }
+}
+
+export class ANTVersionsReadable implements AoANTVersionsRead {
+  protected process: AOProcess;
+  constructor(config?: ProcessConfiguration) {
+    if (config === undefined || Object.keys(config).length === 0) {
+      this.process = new AOProcess({
+        processId: ANT_REGISTRY_ID,
+      });
+    } else if (isProcessConfiguration(config)) {
+      this.process = config.process;
+    } else if (isProcessIdConfiguration(config)) {
+      this.process = new AOProcess({
+        processId: config.processId,
+      });
+    } else {
+      throw new InvalidContractConfigurationError();
+    }
+  }
+  async getANTVersions(): Promise<
+    Record<string, { moduleId: string; luaSourceId?: string; notes: string }>
+  > {
+    const res = await this.process.read<
+      Record<string, { moduleId: string; luaSourceId?: string; notes: string }>
+    >({
+      tags: [{ name: 'Action', value: 'Get-Versions' }],
+    });
+    return Object.fromEntries(
+      Object.entries(res).sort(([a], [b]) => a.localeCompare(b)),
+    );
+  }
+
+  async getLatestANTVersion(): Promise<{
+    version: string;
+    moduleId: string;
+    luaSourceId?: string;
+    notes?: string;
+  }> {
+    const versions = await this.getANTVersions();
+    const lastestVersion = Object.entries(versions).at(-1);
+    if (!lastestVersion) throw new Error('No version found');
+    return {
+      version: lastestVersion[0],
+      ...lastestVersion[1],
+    };
+  }
+}
+
+export class ANTVersionsWritable
+  extends ANTVersionsReadable
+  implements AoANTVersionsWrite
+{
+  private signer: AoSigner;
+
+  constructor({ signer, ...config }: WithSigner<ProcessConfiguration>) {
+    super(config);
+    this.signer = createAoSigner(signer);
+  }
+  async addVersion(params: {
+    version: string;
+    moduleId: string;
+    luaSourceId?: string;
+    notes?: string;
+  }) {
+    return this.process.send({
+      tags: pruneTags([
```

Comment:
i feel like these should all be required

Reply:
The contract takes them as optional so i left them optional here